tests/Unit/Helpers/Evolution/DiscontinuousGalerkin/BoundaryConditions.hpp

```diff
@@ -0,0 +1,624 @@
+ *
+ * The boundary conditions return a `std::optional<std::string>` that is the
+ * error message. For ghost cell boundary conditions they must also return the
+ * arguments needed by the boundary corrections `dg_package_data` function by
+ * `gsl::not_null`. Time derivative boundary conditions return the correction
+ * added to the time derivatives by `gsl::not_null`, while outflow boundary
+ * conditions should only check that the boundary is actually an outflow
+ * boundary. Therefore, the comparison implementation in python must have a
+ * function for each of these. Which function is called is specified using a
+ * `python_boundary_condition_functions` and `Tags::PythonFunctionName`.
+ *
+ * The specific boundary condition and system need to be explicitly given.
+ * Once the boundary condition and boundary correction base classes are
+ * listed/available in the `System` class we can remove needing to specify
+ * those. The `ConversionClassList` is forwarded to `pypp` to allow custom
+ * conversions of classes to python, such as analytic solutions or equations of
+ * state.
+ *
+ * - The random number generator is passed in so that the seed can be easily
+ *   controlled externally. Use, e.g. `MAKE_GENERATOR(gen)` to create a
+ *   generator.
+ * - The python function names are given in a `TaggedTuple`. A
+ *   `TestHelpers::evolution::dg::Tags::PythonFunctionForErrorMessage` tag must
+ *   be given for specifying the error message that the boundary condition
+ *   returns. In many cases this function will simply return `None`. The tag can
+ *   optionally specify the boundary correction for which it is to be used, so
+ *   that a different error message could be printed if a different boundary
+ *   correction is used.
+ *   The `TestHelpers::evolution::dg::Tags::PythonFunctionName` tag is used to
+ *   give the name of the python function for each return argument. The tags are
+ *   the input to the `package_data` function for Ghost boundary conditions, and
+ *   `::Tags::dt<evolved_var_tag>` for time derivative boundary conditions.
+ * - `factory_string` is a string used to create the boundary condition from the
+ *   factory
+ * - `face_points` is the grid points on the interface. Generally 5 grid points
+ *   per dimension in 2d and 3d is recommended to catch indexing errors. In 1d
+ *   there is only ever one point on the interface
+ * - `box_of_volume_data` is a `db::DataBox` that contains all of the
+ *   `dg_gridless_tags` of the boundary condition. This is not a `TaggedTuple`
+ *   so that all the different types of tags, like base tags, can be supported
+ *   and easily tested.
+ * - `ranges` is a `TaggedTuple` of
+ *   `TestHelpers::evolution::dg::Tags::Range<tag>` specifying a custom range in
+ *   which to generate the random values. This can be used for ensuring that
+ *   positive quantities are randomly generated on the interval
+ *   `(0,upper_bound]`. The default interval if a tag is not listed is `[-1,1)`.
```

Comment:
1. why pass in a range if only the upper bound is used? Or should the documentation say 0 -> lower_bound?
2. the first interval is (] but the second is [). Is one of these a typo?